hs_tracking/models.py

```diff
@@ -0,0 +1,147 @@
+        session = Session.objects.create(visitor=visitor)
+        request.session['hs_tracking_id'] = signing.dumps({'id': session.id})
+        return session
+
+
+class Visitor(models.Model):
+    first_seen = models.DateTimeField(auto_now_add=True)
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, null=True, blank=True, unique=True)
+
+    def export_visitor_information(self):
+        """Exports visitor profile information."""
+        info = {
+            "id": self.id,
+        }
+        if self.user:
+            profile = UserProfile.objects.get(user=self.user)
+            for field in PROFILE_FIELDS:
+                info[field] = getattr(profile, field)
+            for field in USER_FIELDS:
+                info[field] = getattr(self.user, field)
+        else:
+            profile = None
+            for field in PROFILE_FIELDS:
+                info[field] = None
+            for field in USER_FIELDS:
+                info[field] = None
+        return info
+
+
+class Session(models.Model):
+    begin = models.DateTimeField(auto_now_add=True)
+    visitor = models.ForeignKey(Visitor)
+
+    objects = SessionManager()
+
+    def get(self, name):
+        return Variable.objects.filter(session=self, name=name).first().get_value()
+
+    def getlist(self, name):
+        return [v.get_value() for v in Variable.objects.filter(session=self, name=name)]
+
+    def record(self, *args, **kwargs):
+        args = (self,) + args
+        return Variable.record(*args, **kwargs)
+
+
+class Variable(models.Model):
+    TYPES = (
+        ('Integer', int),
+        ('Floating Point', float),
+        ('Text', unicode),
+        ('Flag', bool),
+    )
+    TYPE_CHOICES = (
+        (i, label)
+        for (i, label) in
+        enumerate((label for (label, coercer) in TYPES), 1)
+    )
+
+    session = models.ForeignKey(Session)
+    timestamp = models.DateTimeField(auto_now_add=True)
+    name = models.CharField(max_length=32)
+    type = models.IntegerField(choices=TYPES)
+    value = models.CharField(max_length=130)
+
+    def get_value(self):
+        try:
+            f = float(self.value)
+        except ValueError:
+            f = None
+        try:
+            i = int(self.value)
+        except ValueError:
+            i = None
+        if i is not None and i == f:
+            return i
+        elif f is not None:
+            return f
+        elif self.value == 'true':
+            return True
+        elif self.value == 'false':
+            return False
+        return self.value
+
+    @classmethod
+    def record(cls, session, name, value):
+        for i, (label, coercer) in enumerate(cls.TYPES, 1):
+            try:
+                if value == coercer(value):
+                    type_code = i
+                    break
+            except ValueError:
+                continue
+        return Variable.objects.create(session=session, name=name, type=type_code, value=cls.encode(value))
```

Comment:
@ironfroggy Although unlikely, I think it is possible the passed in value is not one of the supported types, in which case, type_code is not set properly but the last return statement is still wrongly executed. I think it is better to initially set type_code to be -1 for example, and only execute the last return statement when type_code is not -1.